File: CSharpFundamentals/CSharpOOPbasics/InheritanceEx/Mankind/Worker.cs
using System;
using System.Text;

namespace Mankind
{
    public class Worker : Human
    {
        private const int MIN_SALARY = 10;
        private const int WORKING_HOURS_MIN = 1;
        private const int WORKING_HOURS_MAX = 12;
        private decimal salary;
        private double workHours;

        public Worker(string firstName, string lastName, decimal salary, double workHour)
            : base(firstName, lastName)
        {
            this.Salary = salary;
            this.WorkHours = workHour;
        }

        public double WorkHours
        {
            get { return workHours; }
            private set
            {
                if (value < WORKING_HOURS_MIN || value > WORKING_HOURS_MAX)
                {
                    throw new ArgumentException("Expected value mismatch! Argument: workHoursPerDay");
                }
                workHours = value;
            }
        }

        public decimal Salary
        {
            get { return salary; }
            private set
            {
                if (value <= MIN_SALARY)
                {
                    throw new ArgumentException("Expected value mismatch! Argument: weekSalary");
                }
                salary = value;
            }
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.AppendLine(base.ToString());
            sb.AppendLine($"Week Salary: {this.Salary:F2}");
            sb.AppendLine($"Hours per day: {this.WorkHours:F2}");
            sb.AppendLine($"Salary per hour: {this.CalcSalary():F2}");
            return sb.ToString().TrimEnd();
        }

        private decimal CalcSalary()
        {
            double totalHours = 5 * this.WorkHours;
            return Salary / (decimal)totalHours;
        }
    }
}

File: JavaDB/DatabaseAdvancedFrameworks/JavaOOPPrinciplesEx/Mankind/models/Worker.java
package Mankind.models;

public class Worker extends Human {
    private static final double MIN_WEEK_SALARY = 10;
    private static final int MIN_WORKING_HOURS = 1;
    private static final int MAX_WORKING_HOURS = 12;
    private static final int MIN_LAST_NAME_LENGTH = 3;
    private double weekSalary;
    private double hoursWork;
    private double salaryPerHour;

    public Worker(String firstName, String lastName, double weekSalary, double hoursWork) {
        super(firstName, lastName);
        this.setWeekSalary(weekSalary);
        this.setHoursWork(hoursWork);
        this.setSalaryPerHour();
    }

    public double getSalaryPerHour() {
        return salaryPerHour;
    }

    public void setSalaryPerHour() {
        this.salaryPerHour = (weekSalary / 7) / this.hoursWork;
    }

    public double getWeekSalary() {
        return weekSalary;
    }

    public void setWeekSalary(double weekSalary) {
        if (weekSalary <= MIN_WEEK_SALARY){
            throw new IllegalArgumentException(
                    "Expected value mismatch!Argument: weekSalary"
            );
        }

        this.weekSalary = weekSalary;
    }

    public double getHoursWork() {
        return hoursWork;
    }

    public void setHoursWork(double hoursWork) {
        if (hoursWork < MIN_WORKING_HOURS || hoursWork > MAX_WORKING_HOURS){
            throw new IllegalArgumentException(
                    "Expected value mismatch!Argument: workHoursPerDay"
            );
        }

        this.hoursWork = hoursWork;
    }

    @Override
    public void setLastName(String lastName) {
        if (Character.isLowerCase(lastName.charAt(0))){
            throw new IllegalArgumentException(
                    "Expected upper case letter!Argument: lastName");
        }

        if (lastName.length() < MIN_LAST_NAME_LENGTH){
            throw new IllegalArgumentException(String.format(
                    "Expected length more than %d symbols!Argument: lastName",
                    MIN_LAST_NAME_LENGTH));
        }

        super.setLastName(lastName);
    }

    @Override
    public String toString(){
        return super.toString() +
                String.format("Week Salary: %.2f%n", this.getWeekSalary()) +
                String.format("Hours per day: %.2f%n", this.getHoursWork()) +
                String.format("Salary per hour: %.2f", this.getSalaryPerHour());
    }
}
